File: Nebula.Interop/src/InterpreterWords.cpp
#include "InterpreterWords.h"
#include "EmitTokens.h"
#include "Utility.h"
#include "InstructionDefs.h"
using namespace nebula::shared;
using namespace nebula;

String^ Nebula::CodeEmitter::InterpreterWords::GetScriptSectionName(ScriptSection type, bool withPrefix)
{
    const char* sectionName = ReadableScriptSection((nebula::ScriptSection)type);
    String^ suffix = gcnew String(sectionName);

    if (withPrefix)
    {
        String^ prefix = GetTokenChar(TokenType::MarkerPrefix);
        return prefix + suffix;
    }

    return suffix;
}

String^ Nebula::CodeEmitter::InterpreterWords::GetTokenChar(TokenType type)
{
    switch (type)
    {
    case TokenType::OpenBody:
        return gcnew String(EMIT_OPENBODY_TOKEN);
    case TokenType::CloseBody:
        return gcnew String(EMIT_CLOSEBODY_TOKEN);
    case TokenType::AttributePrefix:
        return gcnew String(ATTRIBUTE_PREFIX);
    case TokenType::OpenParenthesis:
        return gcnew String(OPEN_PARENTHESIS);
    case TokenType::ClosedParenthesis:
        return gcnew String(CLOSE_PARENTHESIS);
    case TokenType::MarkerPrefix:
        return gcnew String(GENERIC_MARKER_PREFIX);
    case TokenType::CompiledComment:
        return gcnew String(COMMENT_PREFIX);
    }

    throw gcnew System::Exception("Unknown marker");
}

File: Nebula.Interop/Enumerators/InterpreterWords.cs
using System;

namespace Nebula.Interop.Enumerators
{
    public static class InterpreterWords
    {
        public static string GetTokenChar(TokenType type)
        {
            switch (type)
            {
                case TokenType.OpenBody:
                    return "{";
                case TokenType.CloseBody:
                    return "}";
                case TokenType.AttributePrefix:
                    return ";";
                case TokenType.OpenParenthesis:
                    return "(";
                case TokenType.ClosedParenthesis:
                    return ")";
                case TokenType.MarkerPrefix:
                    return ".";
                case TokenType.CompiledComment:
                    return "#";
            }

            throw new Exception("Unknown marker");
        }

        public static string GetScriptSectionName(ScriptSection section, bool withPrefix)
        {
            string sectionString = section.ToString().ToLower();
            switch (section)
            {
                case ScriptSection.Function:
                    sectionString = "func";
                    break;
            }

            if (withPrefix)
            {
                sectionString = GetTokenChar(TokenType.MarkerPrefix) + sectionString;
            }

            return sectionString;
        }
    }
}
